[ui] Add unit tests used by new volume page

ChengYanJin opened the pull request:
**Component**: ui, test

**Context**: Add Unit test for Utility Functions


**Summary**:
- `addMissingDataPoint()` used by Metrics Chart to maually add the missing data points.
- `getVolumeListData()` used by VolumeListTable to retrive the data for volume table

Remove the function `jointDataPointBaseonTimeSeries()` which merges the result of the query based on `instance` and `device` since the Prometheus query can help us handle the aggregation.


Already exist:
Delete a Non-deletable volume and check if UI prevents the users to do so.

==========
Note that: 
Update the storageclass from `metalk8s-prometheus` to `metalk8s` due to the change of PR #2755.

**Acceptance criteria**: 
Pass all the tests

---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #2747 

<!-- If you want to refer to an issue while not closing it, use:

See: #ISSUE_NUMBER

-->

ChengYanJin committed:
ui/tests: Add unit test for addMissingDataPoint functions

Refs:2747

ChengYanJin committed:
ui/tests: Add unit test for getVolumeListData function

Refs: #2747

bert-e commented:
# Hello chengyanjin,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @alexandre-allard-scality

* @tcarmet

* @nootal

* @teddyandrieux

* @lucieleonard

* @gdemonet

* @sayf-eddine-scality

* @nicolast

* @jbertran

* @monpote

* @carlito-scality

* @slaperche-scality

* @cmonfort

ChengYanJin committed:
ui/volumes: Remove the utility function jointDataPointBaseonTimeSeries

Since we introduce the aggregate sum function in prometheus query, we don't need to manually merge the values

Refs: #2747

ChengYanJin committed:
ui/volumes: Add props checking for addMissingDataPoint function

Refs: #2747

ChengYanJin committed:
ui/volumes: Aggregate the result of Prometheus queries by `instance` and `device`

- The sum() function provide by Prometheus can group the result
- Use volume capacity return by prometheus to compute the current volume usage
  same for volume usage metrics graph, will address it in the coming PR

Refs: #2747

ChengYanJin committed:
ui/tests: Handle the case when the node filter is not valid in the URL

Refs: #2747

ChengYanJin committed:
ui/tests: Handle the case when the `sampleDuration` and `sampleFrequency` is not valid

We have three predefined the durations and frequencies according to the timespan
If the duration and frequency are out of the list, we should return undefined

Refs: #2747

ChengYanJin committed:
ui/volumes: Return empty array instead of undefined in addMissingDataPoint()

Refs: #2747

ChengYanJin committed:
ui/volumes: Add no data text for the graphs when there is data available

Refs: #2747

ChengYanJin committed:
ui/tests: Update the storage class from `metalk8s-prometheus` to ` metalk8s`

Due to the change of PR #2755, we want to update the storage class

Refs: #2747

MonPote commented on ui/src/services/utils.js:
```diff
@@ -303,8 +282,17 @@ export function addMissingDataPoint(
+    !orginalValues ||
+    orginalValues.length === 0 ||
+    !startingTimeStamp ||
+    !sampleDuration ||
+    !sampleFrequency ||
```
This can be removed since it's already handle by the next 2 lines.

MonPote commented on ui/src/components/VolumeMetricGraphCard.js:
```diff
@@ -216,7 +218,8 @@ const MetricGraphCard = (props) => {
   );
   const volumeIOPSData = volumeIOPSWriteData?.concat(volumeIOPSReadData);
-
+  console.log('volumeIOPSData', volumeIOPSData);
+  console.log('volumeThroughputData', volumeThroughputData);
```
This can be remove ^^

gdemonet commented on ui/src/ducks/app/monitoring.js:
```diff
@@ -330,15 +334,20 @@ export function* fetchVolumeStats() {
-  const volumeIOPSReadQuery = `irate(node_disk_reads_completed_total{job="node-exporter"}[5m])`;
-  const volumeIOPSWriteQuery = `irate(node_disk_writes_completed_total{job="node-exporter"}[5m])`;
-  const volumeUsedQuery = 'kubelet_volume_stats_used_bytes';
+  // group the result of the query by instance and device (remove the filter {job="node-exporter"})
```
https://github.com/scality/metalk8s/pull/2757#discussion_r477327539

ChengYanJin replied:
Fixed in #2757

gdemonet commented on ui/src/ducks/app/monitoring.js:
```diff
@@ -427,28 +436,40 @@ export function* fetchVolumeStats() {
+  const volumeLatencyCurrentQuery = `irate(node_disk_io_time_seconds_total{job="node-exporter"}[1h])`;
   const volumeUsedQuery = 'kubelet_volume_stats_used_bytes';
-
+  // Grafana - Used Space: kubelet_volume_stats_capacity_bytes - kubelet_volume_stats_available_bytes
```
IIUC, this comment relates to the `volumeUsedQuery` above - in general, we write comments above the related code, not under it

gdemonet commented on ui/src/ducks/app/monitoring.js:
```diff
@@ -330,15 +334,20 @@ export function* fetchVolumeStats() {
+
+  // the query for latency
+  // Disk latency is the time that it takes to complete a single I/O operation on a block device
+  // rate(node_disk_read_time_seconds_total[5m]) / rate(node_disk_reads_completed_total[5m]
+  const volumeLatencyQuery = `sum(irate(node_disk_io_time_seconds_total[1m])) by (instance, device)`;
```
Not sure I understand this query... The commented query makes more sense to me, where did you get the other one from?

ChengYanJin replied:
Yes! the commented query is the correct one actually. This is exactly the bug that #2757 fixed.

gdemonet commented on ui/src/ducks/app/monitoring.js:
```diff
@@ -427,28 +436,40 @@ export function* fetchVolumeStats() {
+    volumeCapacityCurrent = volumeCapacityCurrentQueryResult.data.result;
+  }
+
   const volumeLantencyCurrentResult = yield call(
```
nitpick: there's a typo here `s/Lantency/Latency`

gdemonet commented on ui/src/services/NodeVolumesUtils.js:
```diff
@@ -223,21 +231,26 @@ export const getVolumeListData = createSelector(
+      const volumeCurrentLatency = volumeLatencyCurrent?.find(
+        (vLV) =>
+          vLV.metric.device === volume?.status?.deviceName &&
+          vLV.metric.instance === instanceIP + PORT_NUMBER_PROMETHEUS,
```
Shouldn't it be:
```suggestion
          vLV.metric.instance === `${instanceIP}:${PORT_NUMBER_PROMETHEUS}`,
```
?

ChengYanJin replied:
correct! It's more understandable!

gdemonet replied:
But it doesn't mean the same thing... So is `PORT_NUMBER_PROMETHEUS` actually a string containing `:9090`?

ChengYanJin replied:
yes.. originally the value of `PORT_NUMBER_PROMETHEUS` is `:9090` 
but yeah, it doesn't make any sense.

gdemonet commented on ui/src/services/NodeVolumesUtils.js:
```diff
@@ -223,21 +231,26 @@ export const getVolumeListData = createSelector(
+                (volumeUsedCurrent?.value[1] / volumeCapacityCurrent.value[1]) *
+                100
+              ).toFixed(2)
+            : intl.translate('unknown'),
```
Unrelated to this change, but it'd make more sense to leave `undefined` as a value here and handle the choice of message and its translation in a display component.

gdemonet commented on ui/src/services/NodeVolumesUtils.js:
```diff
@@ -252,12 +265,7 @@ export const getVolumeListData = createSelector(
-          )
+          volumeCurrentLatency
             ? Math.round(
                 volumeLatencyCurrent?.find(
```
You already found the latency for this volume, no need to `.find` again, just re-use `volumeCurrentLatency`

gdemonet commented on ui/src/services/utils.test.js:
```diff
@@ -91,14 +95,559 @@ it('test the sort with a custom sortBy', () => {
   ]);
 });
+
+const originalValue_happy_path = [
```
nit: why mix camel-case with snake-case?

gdemonet commented on ui/src/services/utils.test.js:
```diff
@@ -91,14 +95,559 @@ it('test the sort with a custom sortBy', () => {
+  [1597814571, '5148012544'],
+  [1597818171, '5303730176'],
+  [1597821771, '5234618368'],
+  [1597825371, '5384785920'],
```
I don't think it's relevant to use 130 datapoints for a unit-test. Just get the minimal amount needed, that will help make the test case more obvious.

gdemonet commented on ui/src/services/utils.test.js:
```diff
@@ -91,14 +95,559 @@ it('test the sort with a custom sortBy', () => {
+  [1597821771, '5234618368'],
+  [1597825371, '5384785920'],
+];
+const startingTimeStamp = 1597220571;
```
For the same reason, you don't need "real" values in a unit test (unless they are compared with real values you cannot control. So, starting from timestamp 0 is OK, in my opinion, and makes it more obvious for a reader to understand what's happening.

gdemonet commented on ui/src/services/utils.test.js:
```diff
@@ -91,14 +95,559 @@ it('test the sort with a custom sortBy', () => {
+    sampleFrequency,
+  );
+  expect(result).toEqual(originalValue_all_0);
+});
```
This is a duplicate of the above.

gdemonet commented on ui/src/services/NodeVolumesUtilsData.js:
```diff
@@ -0,0 +1,2371 @@
+export const state_volume_filtered_by_node_master0 = {
```
How is this variable describing "volumes filtered by node master0"? It looks to me as a rather "normal" application state, with three nodes and volumes on each node...

gdemonet commented on ui/src/services/NodeVolumesUtils.test.js:
```diff
@@ -420,3 +422,96 @@ it('should return error when called with failed Ready condition', () => {
+    isExact: false,
+    params: {},
+  },
+};
```
Do we really need all these props? It seems we only read the `location` part anyway...

On the side, it seems to me there are way too many things happening in this selector, and maintaining this code and associated tests will be a pain when the behaviour changes. We'll see.

gdemonet commented on ui/src/services/NodeVolumesUtils.js:
```diff
@@ -168,13 +168,26 @@ export const getVolumeListData = createSelector(
+    let nodeVolumes = null;
+
+    if (nodeFilter) {
+      // when there is node filter in URL, check if the node name is valid / exist
+      const node = nodeList?.find((node) => node.name === nodeFilter);
+      if (!node) {
+        return [];
+      } else {
+        // the node name exists, check if there is any volume belong to this node
+        nodeVolumes = volumes?.filter(
+          (volume) => volume.spec.nodeName === nodeFilter,
+        );
+        if (!nodeVolumes) {
+          return [];
+        }
+      }
```
That's excessively complex IMO, unless we really want to prevent the case where someone inputs a wrong node name both in the volume spec and in the URL... Otherwise, this would suffice:
```suggestion
      nodeVolumes = volumes?.filter(
        (volume) => volume.spec.nodeName === nodeFilter,
      );
```

gdemonet commented on ui/src/services/NodeVolumesUtils.js:
```diff
@@ -168,13 +168,26 @@ export const getVolumeListData = createSelector(
+    let nodeVolumes = null;
+
+    if (nodeFilter) {
+      // when there is node filter in URL, check if the node name is valid / exist
```
If this case is really important to consider, shouldn't we treat it more explicitly for the user? Otherwise, from a volume listing standpoint, it is the same to filter by a node which has no volume, or a non-existing node (since we just return an empty list).

ChengYanJin replied:
Absolutely right! 
Since we return an empty array for both cases, a node which has no volume and a non-existing node, I shouldn't implement such a complex logic here!

gdemonet commented on ui/src/services/utils.js:
```diff
@@ -285,7 +287,10 @@ export function addMissingDataPoint(
     !sampleDuration ||
-    !sampleFrequency
+    !sampleFrequency ||
+    // if the `sampleDuration` and `sampleFrequency` is not from the predefined value
+    !SAMPLE_DURATION_LIST.includes(sampleDuration) ||
+    !SAMPLE_FREQUENCY_LIST.includes(sampleFrequency)
```
Why add this behaviour? It only seems an artificial limitation to me (not showing an option in the visible UI doesn't mean your utility function can't do something).

ChengYanJin replied:
For me, this function is **only** useful for the graph. And the choices of `sampleDuration` and `sampleFrequency` are quite limited (depends on the possible timespan). 
I add this check is to prevent the wrong inputs.

Does it make sense?

gdemonet replied:
> For me, this function is **only** useful for the graph.

Mmmh. To me, this function is only **used** by the current charts we have, which only have some particular values as input. However, it could be useful for other charts using other sample sizes/steps, no?

> And the choices of sampleDuration and sampleFrequency are quite limited (depends on the possible timespan).
> I add this check is to prevent the wrong inputs.

I understand that we only give a user certain choices, through the selection of a timespan. But why would another choice be **wrong** for this utility function?

Let's take another example, coming from the real world. Imagine, a common tool called a screwdriver. Imagine you're building a screwdriver, so you can provide it to your users for, you guessed it... driving some screws. Now, let's say you only give 30mm, 40mm, and 50mm screws to your users, to use with your screwdriver. Does it mean the screwdriver **cannot** drive 60mm screws? Should the screwdriver break if it were given such a screw, by design?

ChengYanJin replied:
Thanks for the explanation! 

I agree! Since it's a utility function we should not limit the choices. 

We can still check the properties `sampleFrequency` and `sampleDuration` if they're based on the choices we provided. But the checking should be at where we call the function.

gdemonet reviewed (changes requested):
Mostly comments around behaviour change and existing code, so you may relegate them to a debt ticket.

However, I think there's some effort to spend on:
- designing utility methods to limit their concern to single, simple tasks
- writing clear, minimal test cases, to make the tested behaviour obvious for a reader

Testing higher-order functions which interact with "real-life" datasets isn't really adapted to unit-testing (which is really about testing complex algorithmic behaviour), but could be worth writing integration tests for.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @alexandre-allard-scality

* @tcarmet

* @nootal

* @teddyandrieux

* @lucieleonard

* @gdemonet

* @sayf-eddine-scality

* @nicolast

* @jbertran

* @monpote

* @carlito-scality

* @slaperche-scality

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @gdemonet

ChengYanJin committed:
ui: Improvement

Refs: #2747

ChengYanJin committed:
ui/test: Remove the properties `sampleFrequency` and `sampleDuration` limitation in utility function

Improve the test set with human-understandable data instead of the real data

ChengYanJin committed:
ui/tests: Addressed the comments for getVolumeListData function

- Simplfied the state, remove the useless props
- Simplfied the logic in getVolumeListData, since we should have the same behaviour for the wrong Node filter and non-volume Node
- Use camelCase for all the mutable variables

ChengYanJin commented:
merged in PR #2779
